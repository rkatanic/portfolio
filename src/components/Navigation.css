.icon {
  height: 1.5rem;
  fill: var(--color-text);
  cursor: pointer;
}

.icon:hover {
  fill: var(--color-primary);
}

.header {
  margin: 2.5rem 0;
  z-index: var(--z-index-100);
}

.nav {
  font-size: 0.825rem;
  font-weight: 600;
  color: var(--color-title);
  max-width: var(--width-desktop);
  display: flex;
  flex-direction: column;
  margin: auto;
  height: 100%;
}

.nav-left {
  display: flex;
  align-items: center;
}

.night-mode-switch {
  display: flex;
}

.nav-list {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  padding: 0;
  margin: 0;
}

.nav-item {
  padding: 1rem 0;
  transition: 0.5s;
  display: flex;
}

.dash {
  width: 4.5rem;
  height: 2px;
  margin: 1rem;
  background-color: var(--color-title);
  transition: 0.5s ease;
}

.number {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: var(--color-input);
  width: 2.25rem;
  height: 2.25rem;
}

.nav-link {
  color: var(--color-title);
  display: flex;
  align-items: center;
  text-transform: uppercase;
  letter-spacing: 2px;
}

.nav-link:hover {
  color: var(--color-primary);
}
.nav-link:hover .dash {
  background-color: var(--color-primary);
}

.active .dash {
  width: 7rem;
  background-color: var(--color-primary);
}

.active a {
  color: var(--color-primary);
}

.nav-toggle {
  display: flex;
  align-items: center;
  justify-content: center;
  position: fixed;
  bottom: 2rem;
  right: 2rem;
  width: 3rem;
  height: 3rem;
  background-color: var(--color-primary);
  z-index: 2;
}

.nav-toggle .icon {
  fill: black;
}

.nav-close {
  position: absolute;
  right: 1.5rem;
  top: 1.5rem;
  height: 1.5rem;
  fill: var(--color-text);
}

.nav-menu {
  height: 100%;
}

@media screen and (max-width: 767px) {
  .nav-menu {
    width: 100%;
    height: initial;
    position: fixed;
    bottom: -100%;
    left: 0;
    padding: 3rem;
    transition: 0.5s;
    background-color: var(--color-body);
    box-sizing: border-box;
  }

  .nav-list {
    align-items: center;
  }

  .nav-item {
    margin: 1rem;
  }

  .dash,
  .number {
    display: none;
  }
}

@media screen and (min-width: 768px) {
  .nav-toggle,
  .nav-close {
    display: none;
  }
}

.show-menu {
  bottom: 0;
  z-index: 3;
}
