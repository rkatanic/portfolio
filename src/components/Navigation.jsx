import { useEffect, useState } from "react";
import { useIsDarkMode, useToggleDarkMode } from "../context/GlobalContext";
import { ReactComponent as MenuIcon } from "../assets/icons/menu.svg";
import { ReactComponent as SunIcon } from "../assets/icons/sun.svg";
import { ReactComponent as MoonIcon } from "../assets/icons/moon.svg";

import "./Navigation.css";

const Navigation = () => {
  const hueColors = [
    { key: "red", value: 5 },
    { key: "green", value: 145 },
    { key: "blue", value: 210 },
    { key: "purple", value: 270 },
  ];

  const [showMenu, setShowMenu] = useState(false);

  const isDarkMode = useIsDarkMode();
  const toggleDarkMode = useToggleDarkMode();

  const shouldShowDarkMode = () => {
    if (isDarkMode) {
      window.document.body.setAttribute("class", "dark-mode");
    } else {
      window.document.body.removeAttribute("class");
    }
  };

  useEffect(() => {
    shouldShowDarkMode();
  }, [isDarkMode]);

  const handleMenuToggle = () => setShowMenu((prevState) => !prevState);

  const handleColorSwitch = (colorValue) => {
    window.document.documentElement.style.setProperty(
      "--color-hue",
      colorValue
    );
  };

  return (
    <header className="header">
      <nav className="nav">
        <div className="nav-left">
          <div className="color-switch-container">
            {hueColors.map((hueColor) => (
              <div
                onClick={() => handleColorSwitch(hueColor.value)}
                key={hueColor.key}
                className="color-indicator"
                style={{
                  backgroundColor: `hsl(${hueColor.value},70%,60%)`,
                  border: `2px solid hsl(${hueColor.value},90%,75%)`,
                }}
              ></div>
            ))}
          </div>
          <div className="night-mode-switch" onClick={toggleDarkMode}>
            {isDarkMode ? <SunIcon /> : <MoonIcon />}
          </div>
        </div>
        <div className={showMenu ? `nav-menu show-menu` : `nav-menu`}>
          <ul className="nav-list">
            <li className="nav-item" onClick={() => setShowMenu(false)}>
              <a href="#home" className="nav-link ">
                Home
              </a>
            </li>
            <li className="nav-item" onClick={() => setShowMenu(false)}>
              <a href="#about" className="nav-link">
                About
              </a>
            </li>
            <li className="nav-item" onClick={() => setShowMenu(false)}>
              <a href="#contact" className="nav-link">
                Contact
              </a>
            </li>
          </ul>
          <div className="nav-close" onClick={handleMenuToggle}>
            x
          </div>
        </div>
        <div className="nav-btns">
          <div className="nav-toggle">
            <MenuIcon onClick={handleMenuToggle} />
          </div>
        </div>
      </nav>
    </header>
  );
};

export default Navigation;
