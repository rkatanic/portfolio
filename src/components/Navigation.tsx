import { useState } from "react";
import { ReactComponent as MenuIcon } from "../assets/icons/menu.svg";
import { ReactComponent as CloseIcon } from "../assets/icons/close.svg";
import Scrollspy from "react-scrollspy";

import "./Navigation.css";

const Navigation = (): JSX.Element => {
  const [showMenu, setShowMenu] = useState(false);

  const handleMenuToggle = (): void => setShowMenu((prevState) => !prevState);

  return (
    <nav className="nav">
      <div
        className={showMenu ? `nav-menu show-menu` : `nav-menu`}
        data-testid="nav-menu"
      >
        <Scrollspy
          offset={-30}
          className="nav-list"
          currentClassName="active"
          items={["about", "skills", "faq", "contact"]}
        >
          <li className="nav-item" onClick={() => setShowMenu(false)}>
            <a href="#about" className="nav-link">
              <div className="number">01</div>
              <div className="dash"></div>
              <span>About</span>
            </a>
          </li>
          <li className="nav-item" onClick={() => setShowMenu(false)}>
            <a href="#skills" className="nav-link">
              <div className="number">02</div>
              <div className="dash"></div>
              <span>Skills</span>
            </a>
          </li>
          <li className="nav-item" onClick={() => setShowMenu(false)}>
            <a href="#faq" className="nav-link">
              <div className="number">03</div>
              <div className="dash"></div>
              <span>FAQ</span>
            </a>
          </li>
          <li className="nav-item" onClick={() => setShowMenu(false)}>
            <a href="#contact" className="nav-link">
              <div className="number">04</div>
              <div className="dash"></div>
              <span>Contact</span>
            </a>
          </li>
        </Scrollspy>
      </div>
      <div
        className={`nav-toggle ${showMenu ? "close" : ""}`}
        onClick={handleMenuToggle}
        data-testid="nav-toggle"
      >
        <div className="nav-toggle-dash"></div>
      </div>
    </nav>
  );
};

export default Navigation;
