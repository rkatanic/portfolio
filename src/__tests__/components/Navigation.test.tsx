import { fireEvent, render } from "@testing-library/react";
import Navigation from "../../components/Navigation";

describe("Navigation", (): void => {
  it("should render", (): void => {
    const { baseElement } = render(<Navigation />);

    expect(baseElement).toMatchSnapshot();
  });

  it("should handle menu open", (): void => {
    const { getByTestId } = render(<Navigation />);
    const navToggle = getByTestId("nav-toggle");
    const navMenu = getByTestId("nav-menu");

    fireEvent.click(navToggle);

    expect(navToggle).toHaveClass("close");
    expect(navMenu).toHaveClass("show-menu");
  });

  it("should close menu when clicked on settings", (): void => {
    const { getByTestId } = render(<Navigation />);
    const navToggle = getByTestId("nav-toggle");
    const navMenu = getByTestId("nav-menu");

    fireEvent.click(getByTestId("nav-settings"));

    expect(navToggle).not.toHaveClass("close");
    expect(navMenu).not.toHaveClass("show-menu");
  });

  it.each(["About", "Skills", "FAQ", "Contact"])(
    "should close menu when clicked on %s navigation link",
    (navLinkText: string): void => {
      const { getByTestId, getByText } = render(<Navigation />);
      const navToggle = getByTestId("nav-toggle");
      const navMenu = getByTestId("nav-menu");

      fireEvent.click(getByText(navLinkText));

      expect(navToggle).not.toHaveClass("close");
      expect(navMenu).not.toHaveClass("show-menu");
    }
  );
});
