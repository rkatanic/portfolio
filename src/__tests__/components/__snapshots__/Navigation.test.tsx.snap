// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render 1`] = `
<body>
  <div>
    <nav
      class="nav"
    >
      <div
        class="nav-menu"
        data-testid="nav-menu"
      >
        <ul
          class="nav-list"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#about"
            >
              <div
                class="number"
              >
                01
              </div>
              <div
                class="dash"
              />
              <span>
                About
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#skills"
            >
              <div
                class="number"
              >
                02
              </div>
              <div
                class="dash"
              />
              <span>
                Skills
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#faq"
            >
              <div
                class="number"
              >
                03
              </div>
              <div
                class="dash"
              />
              <span>
                FAQ
              </span>
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link"
              href="#contact"
            >
              <div
                class="number"
              >
                04
              </div>
              <div
                class="dash"
              />
              <span>
                Contact
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="nav-toggle "
        data-testid="nav-toggle"
      >
        <div
          class="nav-toggle-dash"
        />
      </div>
    </nav>
  </div>
</body>
`;
